feat: Add nullable vector support in Core Segment, Index, and Query layers

marcelo-cjl opened the pull request:
This commit adds nullable vector support for all vector field types in the core C++ layer including:
- Segment layer (ChunkedSegmentSealedImpl, SegmentGrowingImpl)
- Index layer (VectorDiskIndex, VectorMemIndex, VecIndexCreator)
- Query layer (SearchOnGrowing, SearchOnSealed, SearchOnIndex)
- MMap layer (ChunkedColumn, ChunkedColumnGroup, ChunkedColumnInterface)
- Common utilities (ChunkWriter, FieldMeta, Schema)

Key changes:
- BuildValidRowIds() to track valid (non-null) row indices
- TransformBitset() to map bitsets between data space and index space
- TransformOffset() to convert search results back to logical offsets
- NullableVectorChunkWriter for compact storage of nullable vectors
- Valid data serialization/deserialization in vector indices

marcelo-cjl commented:
/ci-rerun-ut-integration

marcelo-cjl commented:
/ci-rerun-ut

foxspy commented on internal/core/src/index/VectorDiskIndex.cpp:
```diff
@@ -120,6 +120,32 @@ VectorDiskAnnIndex<T>::Load(milvus::tracer::TraceContext ctx,
+        local_chunk_manager->Read(row_ids_path, 0, row_ids_.data(), file_size);
+    }
+
+    auto valid_data_path = local_index_path_prefix + "/valid_data";
```
same here

marcelo-cjl replied:
done

foxspy commented on internal/core/src/index/VectorDiskIndex.cpp:
```diff
@@ -298,6 +324,30 @@ VectorDiskAnnIndex<T>::BuildWithDataset(const DatasetPtr& dataset,
+    }
+
+    if (!valid_data_.empty()) {
+        auto valid_data_path = local_index_path_prefix + "/valid_data";
```
same as above

foxspy commented on internal/core/src/index/VectorIndex.h:
```diff
@@ -145,6 +145,57 @@ class VectorIndex : public IndexBase {
+    }
+
+ protected:
+    FixedVector<int64_t> row_ids_;
```
valid_ids_

foxspy commented on internal/core/src/index/VectorMemIndex.cpp:
```diff
@@ -160,6 +192,33 @@ VectorMemIndex<T>::LoadWithoutAssemble(const BinarySet& binary_set,
+        row_ids_.resize(count);
+        std::memcpy(row_ids_.data(), ptr->data.get(), count * sizeof(int64_t));
+    }
+
```
This is a keyword — please define this key within the Knowhere parameters configuration.

foxspy commented on internal/core/src/indexbuilder/index_c.cpp:
```diff
@@ -374,13 +374,6 @@ BuildJsonKeyIndex(ProtoLayoutInterface result,
                 build_index_info->storage_plugin_context().encryption_key());
-
-            auto plugin_context = std::make_shared<CPluginContext>();
-            plugin_context->ez_id =
```
Why was this removed?

marcelo-cjl replied:
has been restored

foxspy commented on internal/core/src/indexbuilder/index_c.cpp:
```diff
@@ -472,12 +465,6 @@ BuildTextIndex(ProtoLayoutInterface result,
                 build_index_info->storage_plugin_context().encryption_zone_id(),
                 build_index_info->storage_plugin_context().collection_id(),
                 build_index_info->storage_plugin_context().encryption_key());
-            auto plugin_context = std::make_shared<CPluginContext>();
```
same

foxspy commented on internal/core/src/query/SearchOnGrowing.cpp:
```diff
@@ -154,6 +153,24 @@ SearchOnGrowing(const segcore::SegmentGrowingImpl& segment,
         // step 3: brute force search where small indexing is unavailable
         auto vec_ptr = record.get_data_base(vecfield_id);
+        bool is_nullable = field.is_nullable();
+        const auto& valid_data = vec_ptr->get_valid_data();
```
These two structures have the same meaning — they can be merged into one.

foxspy commented on internal/core/src/segcore/reduce/StreamReduce.cpp:
```diff
@@ -98,6 +98,34 @@ StreamReducerHelper::AssembleMergedResult() {
                         merge_output_data_bases[nq_base_offset + loc] = {
                             &search_result->output_fields_data_, ki};
+
+                        for (auto field_id : plan_->target_entries_) {
```
Duplicate code

marcelo-cjl replied:
done

marcelo-cjl committed:
Add nullable support for all vector field types

Supported types: FloatVector, BinaryVector, Float16Vector, BFloat16Vector,
Int8Vector and SparseFloatVector.

C++ layer changes:
- Add FieldDataVectorImpl class with LogicalToPhysicalMapping for sparse storage
- LogicalToPhysicalMapping uses adaptive strategy: map when valid_ratio < 10%,
  otherwise uses vector for O(1) lookup
- Override FillFieldData() in FieldDataVectorImpl to handle nullable vectors
- Move FieldDataSparseVectorImpl from FieldDataInterface.h to FieldData.h
- Remove "vector not support null" restriction in Util.cpp

Go layer changes:
- Add LogicalToPhysicalMapping struct with validCount and l2pMap
- Add ValidData, Nullable, L2PMapping fields to all vector structs
- Update GetRow() to use L2PMapping.GetPhysicalOffset() for physical index
- Update AppendRow()/AppendValidDataRows() to build mapping incrementally
- Update GetMemorySize() to include ValidData, Nullable and L2PMapping

Storage strategy: nullable vectors use sparse storage where Data array only
contains non-null vectors, ValidData bitmap tracks null positions,
L2PMapping translates logical offset to physical offset.

Signed-off-by: marcelo-cjl <marcelo.chen@zilliz.com>

marcelo-cjl committed:
Adapt storage layer for nullable vector support

This commit adapts the storage layer (binlog read/write) to correctly handle
nullable vectors with sparse storage, where Data array only contains valid
vectors, ValidData bitmap tracks null positions, and L2PMapping translates
logical offset to physical offset.

C++ storage layer changes:
- Update PayloadReader to extract dim from Arrow schema metadata for nullable
  vectors and relax IsFull() check for nullable vector fields
- Update PayloadWriter to use BinaryBuilder for nullable vectors with dim
  stored in Arrow metadata
- Update Util add_vector_payload() to handle nullable vectors via BinaryBuilder
- Add FieldDataSparseVectorImpl using declaration to access base FillFieldData
- Update DataCodecTest and DiskFileManagerTest with nullable vector test cases

Go storage layer changes:
- Update PayloadReader to handle nullable vectors reading with sparse data
- Update PayloadWriter to serialize nullable vectors with validity bitmap
- Update serde.go serialization to iterate by logical count (len(ValidData))
  and use GetRow(j) with logical index for nullable vectors
- Update data_codec.go binlog deserialization for nullable vector fields
- Update insert_data.go RowNum() to return logical count (len(ValidData))
  for nullable vectors
- Update utils.go to remove obsolete NumRows field references
- Add comprehensive test coverage for nullable vector payload operations

Storage format: nullable vectors stored as Arrow Binary with dim in metadata,
validity bitmap indicates null positions, data contains only valid vectors.

Signed-off-by: marcelo-cjl <marcelo.chen@zilliz.com>

marcelo-cjl committed:
Add nullable vector support for segment search and indexing

This commit adds nullable vector support for both growing and sealed segments,
including search, retrieval, and index building operations. It introduces
OffsetMapping for efficient logical-to-physical offset translation.

Core infrastructure changes:
- Add OffsetMapping class with Build()/BuildIncremental() for bitmap-based
  offset mapping, supporting both dense lookup and efficient valid count tracking
- Add TransformBitset() and TransformOffset() utilities for converting between
  logical and physical coordinate spaces during search operations
- Update SearchResult to include has_raw_data_ flag and physical offset handling

Growing segment changes:
- Update ConcurrentVector to support nullable vectors with OffsetMapping
- Add get_offset_mapping() and valid data tracking in AckSeal()
- Update FieldIndexing to handle nullable vectors with physical offset storage
- Modify SearchOnGrowing to transform bitset/offsets for nullable vector search
- Add nullable vector tests in SegmentGrowingTest covering search and retrieval

Sealed segment changes:
- Update ChunkedColumn/ChunkedColumnGroup to support OffsetMapping
- Modify ChunkedSegmentSealedImpl to handle nullable vector field data loading
- Update SearchOnSealedIndex/SearchOnSealedColumn with early return for 100% null
  case to prevent crash when searching empty index
- Add FilterVectorValidOffsets() for output field retrieval with nullable vectors
- Expand ChunkedSegmentSealedBinlogIndexTest with comprehensive nullable test cases

Index building changes:
- Update VectorMemIndex/VectorDiskIndex to store and serialize OffsetMapping
- Add BuildValidData()/UpdateValidData() for nullable vector index building
- Update VecIndexCreator and index_c API with valid data parameter support
- Modify InterimSealedIndexTranslator to handle nullable vector binlog index

Search and reduce changes:
- Update Reduce.cpp to transform physical offsets back to logical for output
- Update StreamReduce with nullable vector offset transformation support
- Add early return in SearchOnSealed when offset_mapping has zero valid count

Signed-off-by: marcelo-cjl <marcelo.chen@zilliz.com>

marcelo-cjl commented:
/ci-rerun-ut

marcelo-cjl commented:
/ci-rerun-e2e-default

marcelo-cjl commented on internal/core/src/common/QueryResult.h:
```diff
@@ -195,6 +199,7 @@ struct VectorIterator {
         heap_;
     bool sealed = false;
     std::vector<int64_t> total_rows_until_chunk_;
+    FixedVector<int64_t> row_ids_;
```
has been changed

marcelo-cjl commented on internal/core/src/segcore/ChunkedSegmentSealedImpl.cpp:
```diff
@@ -397,46 +393,6 @@ ChunkedSegmentSealedImpl::LoadColumnGroup(
-
-    auto& mmap_config = storage::MmapManager::GetInstance().GetMmapConfig();
-    bool global_use_mmap = is_vector ? mmap_config.GetVectorFieldEnableMmap()
-                                     : mmap_config.GetScalarFieldEnableMmap();
```
has been restored

marcelo-cjl commented on internal/core/src/index/VectorMemIndex.cpp:
```diff
@@ -160,6 +192,33 @@ VectorMemIndex<T>::LoadWithoutAssemble(const BinarySet& binary_set,
                   "failed to Deserialize index: {}",
                   KnowhereStatusString(stat));
+
+    if (binary_set.Contains("row_ids_count") &&
```
done

marcelo-cjl commented on internal/core/src/index/VectorDiskIndex.cpp:
```diff
@@ -120,6 +120,32 @@ VectorDiskAnnIndex<T>::Load(milvus::tracer::TraceContext ctx,
+    auto local_chunk_manager =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    auto local_index_path_prefix = file_manager_->GetLocalIndexObjectPrefix();
+    auto row_ids_path = local_index_path_prefix + "/row_ids";
```
done

foxspy commented on internal/core/src/index/VectorMemIndex.cpp:
```diff
@@ -146,6 +146,27 @@ VectorMemIndex<T>::Serialize(const Config& config) {
+                data[i / 8] |= (1 << (i % 8));
+            }
+        }
+        ret.Append("valid_data_", data, byte_size);
```
do not use magic key

foxspy reviewed (commented):
/lgtm

marcelo-cjl commented:
feature: #45993
